# Running KEGG Decoder

### References
- [Video tutorial](https://youtu.be/1v4UzjE7K2g?t=962)
- [KEGG](https://www.genome.jp/kegg/)
- [KEGG Decoder Pathway Definitions](https://github.com/bjtully/BioData/blob/master/KEGGDecoder/KOALA_definitions.txt)

### Downloading reference genomes from NCBI (or from any other FTP)
`wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/018/465/GCF_000018465.1_ASM1846v1/GCF_000018465.1_ASM1846v1_genomic.fna.gz`

`wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/015/665/GCF_000015665.1_ASM1566v1/GCF_000015665.1_ASM1566v1_genomic.fna.gz`

`wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/273/775/GCF_001273775.1_ASM127377v1/GCF_001273775.1_ASM127377v1_genomic.fna.gz`

`wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/028/281/495/GCA_028281495.1_ASM2828149v1/GCA_028281495.1_ASM2828149v1_genomic.fna.gz`

`wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/011/805/GCF_000011805.1_ASM1180v1/GCF_000011805.1_ASM1180v1_genomic.fna.gz`

### Decompress files if necessary
`gunzip *gz`

### Rename the sequence identifiers so they comply with Kegg-Decoder naming scheme (i.e. Genome ID is everything that comes before the first "_" character. The sequence IDs should NOT include characters "-", "_", or spaces)
- Note: you might need to change "fna" in the command below to match the extension of your own files (e.g. fasta, or fas)
 
`for genome in GCF_000011805*fna; do id=$(echo $genome | sed -e s'/_\+\|\W\+/./g' ); echo "Changing sequence IDs in" $genome "  to: " $id ; sed -i s"/>/>$id /g" "$genome" ; done`

### Call protein encoding genes with prodigal
`module load prodigal`

`for genome in *fna; do prodigal -q -p single -a "CDS_${genome%.fna}.faa" -d "Genes_${genome%.fna}.fna" -f gff -i $genome -o "${genome%.fna}.gff" ; done`

### Concatenate the protein sequence files into a single file

`cat *faa > All_CDS.faa`

### Query proteins against the KOfam database using hmmsearch
`module load hmmer`

`hmmsearch --noali --tblout All_CDSxKOfam.tsv -o All_CDSxKOfam -E 0.001 --cpu 4 /mnt/smart/scratch/vir/felipe/Databases/KEGG/All_KOs.hmm All_CDS.faa`

### Edit the output to have only the two columns used by KEGG-Decoder
`grep "ID=" All_CDSxKOfam.tsv | cut -d "-" -f 1,2 | sed 's/ //g' |  sed 's/-/\t/g'  > Parsed_All_CDSxKOfam.tsv`

### If not installed already, create and activate a keggdecoder enviroment with conda
`conda create --prefix /mnt/smart/scratch/vir/felipe/envs/keggdecoder python=3.6`

`conda activate /mnt/smart/scratch/vir/felipe/envs/keggdecoder`

`python3 -m pip install KEGGDecoder`

### Run kegg decoder with the output generated by hmmsearch
`conda activate /mnt/smart/scratch/vir/felipe/envs/keggdecoder`

`KEGG-decoder --input Parsed_All_CDSxKOfam.tsv --output All_CDSxKOfam_Decoder.tsv --vizoption static`

`conda deactivate`

### Visualize the results with your preferred tool
`module load R`

`Rscript Make_Metabolic_Heatmap.R`